### PR TITLE
Secure vs insecure image pruning

### LINKED ISSUE #4232: pruning images - note the need for `--certificate-authority` flag when the registry is secured
Emphasize that the `oadm prune images` need to be passed `--certificate-authority` flag if the registry is secured and the current user's authority doesn't match the one used for the secured registry.

Don't forget to mention, what error can be seen if this flag is omitted. List it among common pruning issues.

### PR DESCRIPTION
**Pull request by miminar**
Document new options related to secure connection to integrated docker registry and a mechanism that decides whether to fall-back to insecure connection.

Resolves #4232
Resolves [bz#1469654](https://bugzilla.redhat.com/show_bug.cgi?id=1469654)

~~Is blocked on openshift/origin#14114?~~ No longer blocked.

### CODE REVIEW

**Comment by miminar** [comment]
@legionus PTAL

**Comment by miminar** [comment]
@soltysh PTAL

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: xref:using-insecure-connection-against-secured-registry[the one below]

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: `certificate-authority`
    `registry-url`

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: `prune` command

s/with error similar to/with an error similar to

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: s/config/configuration file
s/-/--

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: I would move the code block to just before "By default...." so that the first sentence does not get broken up.

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: registry, but

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Perhaps we can use [discrete headings ](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#discrete-headings)here instead?

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Perhaps we can use a [discrete heading](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#discrete-headings)here instead?

File: admin_guide/pruning_resources.adoc
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: I'd move the last sentence as a NOTE below. Additionally, reword this to something like:

Whenever possible use `--certificate-authority`, instead. Use of this option is strongly discouraged.

File: admin_guide/pruning_resources.adoc
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: Don't encourage using `--force-insecure` anywhere, so drop the last sentence. It's already explained and that's it, it should not be used anywhere, though.

File: admin_guide/pruning_resources.adoc
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: Suggestion:
The secure connection is the preferred and recommended approach.

It's the recommended as a rule, not only for production.

File: admin_guide/pruning_resources.adoc
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: s/following/the following

Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: following cases **,** unless

File: admin_guide/pruning_resources.adoc
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: Add info in parenthesis this is not recommended.

File: admin_guide/pruning_resources.adoc
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: Add not recommended at the end in parens.

**Review by soltysh — CHANGES_REQUESTED** [review]
I left you some comments.

**Comment by miminar** [comment]
Now blocked on openshift/origin#14405. I will need to update error messages once it lands.

**Comment by miminar** [comment]
New blocked on openshift/origin#14914

File: admin_guide/pruning_resources.adoc
Comment (miminar) [inline admin_guide/pruning_resources.adoc]: I've ditched it. There is plenty of other places advocating the same. Instead, I've put *(Dangerous)* at the beginning.

**Comment by miminar** [comment]
Thanks for all the comments. They should be addressed now. The dependency PR is still waiting for the unblocked merge queue.

**Comment by miminar** [comment]
No longer blocked. @ahardin-rh could you please review once more?

**Comment by bparees** [comment]
@miminar if we label this PR for 3.5 and 3.6, the changes will be applied there.
 
if 3.5 and 3.6 need a *different* set of doc, then you'll need to open a separate PR and only label it for 3.5 and 3.6.

**Comment by bparees** [comment]
@miminar so which versions is this specific PR appropriate for?

**Comment by miminar** [comment]
@bparees 3.6 - latest; I'll re-submit new PRs for 3.4 and 3.5

**Comment by bparees** [comment]
@miminar ok.

@openshift/team-documentation ptal/merge.

**Comment by miminar** [comment]
I did some small corrections to the descriptions of flags.

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Instead of labeling this Dangerous, maybe we can say:

    *Use caution with this option.* Allow...

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: s/below/Using Secure or Insecure Connections

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Using Secure or Insecure Connections

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: is the preferred and recommended approach

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: connection, which is dangerous

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: run with the

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: The provided `registry-url` is prefixed with the `http://` scheme.

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: The provided `registry-url`

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: The configuration of the current user allows for an insecure connection.

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: using the `--certificate-authority` flag.

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: similar to those seen in xref:using-wrong-certificate-authority[Using the Wrong Certificate Authority] or xref:using-insecure-connection-against-secured-registry[Using an Insecure Connection Against a Secured Registry].

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Images Not Being Pruned

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Using a Secure Connection Against an Insecure Registry

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: If you see a message similar to the following in the output of the `oadm prune images` command, then your registry is not secured and the `oadm prune images` client will attempt to use a secure connection:

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: you can force the client to use an insecure

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Using an Insecure Connection Against a Secured Registry

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: use the

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Using the Wrong Certificate Authority

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: that the certificate

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: different than

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: can be added instead

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Docker registry

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Docker registry

File: admin_guide/pruning_resources.adoc
Comment (ahardin-rh) [reply admin_guide/pruning_resources.adoc]: Docker registry

**Comment by ahardin-rh** [comment]
@miminar Thanks! Just a few minor comments from me. 

Just a heads-up that, given our new docs workflow, if you want to submit separate PRs for 3.4 and 3.5, be sure to do so against enterprise-3.4-stage and enterprise-3.5-stage respectively. Thanks again!

**Comment by miminar** [comment]
@ahardin-rh thanks a lot! Comments should be addressed now.

**Comment by ahardin-rh** [comment]
@miminar Excellent. Thank you!

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../admin_guide/pruning_resources.adoc#admin-guide-pruning-resources[Pruning Objects]
|Added details on secure versus insecure image pruning.
%

**Pull request merged**
